### PR TITLE
Add hooks, add tests, improve functionality, refactor documentation, and add GitHub Actions

### PR DESCRIPTION
**Pull request by ESKYoung**
# Summary

Added the following hooks to the project:

- [`black`](https://black.readthedocs.io/en/stable/) - automatically format code. Reduces discrepancies between contributors' formatting. As part of this, refactored all code and documentation
- [`isort`](https://pycqa.github.io/isort/) - automatically sort Python imports
- [`nbQA`](https://nbqa.readthedocs.io/en/latest/index.html) - run QA tools on Jupyter notebooks
   - `black` and `isort` have been enabled, but `flake8` has been disabled until you can skip cells easily

Added more tests to check the template build - these template build tests focus on a TDD-approach to completeness, e.g. checking for counts of Jinja replacements.

Improved functionality including:

- Bug fixing `.envrc`
- Removing local tables of contents (not rendering well in Sphinx)
- Modifying references to indicate public sector usage
- Adding Python and R requirements
- Adding links to the [YouTube presentation](https://www.youtube.com/watch?v=N7_d3k3uQ_M)
- Consolidation of `coverage` and `pytest` configuration files into a `pyproject.toml` file
- Minor fixes for Windows compatibility, including references indicating support for Windows users
- Adding a data log to the `aqa` folder
- Adding a PR template for `govcookiecutter`
- Streamlining and clarifying pre-commit hooks, removed unnecessary hooks at the parent level (e.g. Jupyter notebook hooks)

Streamlined documentation in the parent repository.

Added GitHub Actions to run all tests and coverage for pushes, and also to test build `govcookiecutter` on pull request. Currently not all Actions run for Windows machines, as there are some problems (see issues #29 and #30).

# Checklists

This pull/merge request meets the following requirements:

- [x] Code runs
- [x] Developments are **secure** and [**ethical**][data-ethics-framework]
- [x] You have made proportionate checks that the code works correctly
- [x] You have tested the build of the template
- [x] Test suite passes
- [x] [Minimum usable documentation][agilemodeling] written in the `docs` folder

Comments have been added below around the incomplete checks.

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
> but flake8 has been disabled until you can skip cells easily

I'll get to work then :) Seems like it's a popular request. Would the "celltags" solution work for you?

File: cookiecutter.json [hunk 1]
```diff
@@ -1,16 +1,16 @@
 {
   "_extensions": ["jinja2_time.TimeExtension"],
 
-  "department_name": "Your HM Government department",
+  "organisation_name": "Your public sector organisation name, e.g. GDS",
```
Comment (maxf) [inline cookiecutter.json:4]: Nitpicking here, but encourage full names. Not everyone knows that, for instance, OS is the Ordnance Survey (but also the Office for Students, to add to the confusion)

**Comment by ESKYoung** [comment]
> I'll get to work then :) Seems like it's a popular request. Would the "celltags" solution work for you?
>
> as an aside, 0.12.0 has a few fixes since 0.10.0, I'd suggested upgrading to that. Hopefully 0.13.0 will enable skipping cells

Ah thanks for the tip @MarcoGorelli! And thanks for the package - it's really useful for what we're trying to achieve here! 🎉 

Yes I believe the celltags solution would work - it'd be a [similar solution to keeping notebook cell outputs](https://github.com/ukgovdatascience/govcookiecutter/blob/main/%7B%7B%20cookiecutter.repo_name%20%7D%7D/docs/contributor_guide/pre_commit_hooks.md#keeping-specific-jupyter-notebook-outputs) in `nbstripout`!

File: hooks/post_gen_project.py [hunk 2]
```diff
@@ -1,101 +1,167 @@
...
+    """Delete files and folders for given file and/or folder paths.
 
     Args:
-        folder: A folder path to be removed.
+        paths: A ``pathlib.Path`` object or string, or list of ``pathlib.Path`` or
+        strings.
 
     Returns:
-        None - removes the `folder` folder.
+        None - deletes the files and/or folders defined in ``paths``.
```
Comment (maxf) [inline hooks/post_gen_project.py:16]: Worth mentioning what happens in case there's an error (rmtree failed, etc)?

File: {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md [hunk 3]
```diff
@@ -55,38 +57,43 @@ Next, audit the baseline that has been generated by running:
 detect-secrets audit .secrets.baseline
 ```
 
-When you run this command, you'll enter an interactive console and be presented with a list of high-entropy string
-and/or anything which _could_ be a secret, and asked to verify whether this is the case. By doing this, the hook will
-be in a position to know if you're later committing any _new_ secrets to the repository, and it will be able to alert
-you accordingly.
+When you run this command, you'll enter an interactive console and be presented with a
+list of high-entropy string and/or anything which _could_ be a secret, and asked to
```
Comment (maxf) [inline {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md:61]: strings

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -95,36 +98,33 @@ The HTMl report can be accessed at `htmlcov/index.html`.
...
-## HM Government department frameworks
+We write our documentation in [MyST Markdown][myst] for use in Sphinx. This is mainly
+stored in the `docs` folder, unless it's more appropriate to store it elsewhere, like
+this file.
 
-HM Government frameworks are stored in the `.govcookiecutter/department_frameworks` folder. If you would like to add
-your own department's framework, follow the [instructions][docs-govcookiecutter-frameworks] in the `README.md` file in
-that folder.
+Further information on how to write Sphinx documentation, and how to build it into a
+searchable website can be found [here][docs-write-sphinx-documentation].
```
Comment (maxf) [inline CONTRIBUTING.md:106]: Avoid 'here' as link text, for accessibility reasons. Some assistive tools put the links at the end of the document and "here" doesn't carry over the original context.

**Review by maxf — APPROVED** [review]
Few very minor comments, otherwise LGTM

File: README.md [hunk 5]
```diff
@@ -1,72 +1,83 @@
...
-> details.
+> ℹ️ Contributors have some additional requirements! Check out the
+> [contributing guidelines][contributing] for further details.
 
 To get started your system should meet the following requirements:
 
-1. Unix-based system (macOS, Linux, ...)
-2. Python 3.5+ installed
+1. Python 3.6.1+ installed
+2. R 4.0.4+ installed (for combined Python and R projects)
```
[the remark this answers is not in the document]
Comment (ESKYoung) [reply README.md]: Added in commit [`8359b4e`](https://github.com/ukgovdatascience/govcookiecutter/pull/31/commits/8359b4e43f8e69f188114ead3ab89ef5c4cf54ec).

File: cookiecutter.json (hunk 1 above)
Comment (ESKYoung) [reply cookiecutter.json]: Good spot, my bad! Added in commit [`16902ea`](https://github.com/ukgovdatascience/govcookiecutter/pull/31/commits/16902ea9fed72343297bc6667dce223114facc7a).

File: hooks/post_gen_project.py (hunk 2 above)
Comment (ESKYoung) [reply hooks/post_gen_project.py:17]: I've added more information in the docstring, but will also test what happens with Windows users who may not have admin privileges (GitHub Actions Windows build by default have these privileges). [Concerns based on this SO post](https://stackoverflow.com/questions/2656322/shutil-rmtree-fails-on-windows-with-access-is-denied). Will report back here (although may raise an issue rather than fix it here).

File: {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md (hunk 3 above)
Comment (ESKYoung) [reply {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md]: Done in commit [`178de00`](https://github.com/ukgovdatascience/govcookiecutter/pull/31/commits/178de0063a87f15b35fed7fbb8052cb81d12dbf9).

File: CONTRIBUTING.md (hunk 4 above)
Comment (ESKYoung) [reply CONTRIBUTING.md]: Good call, made the changes (and others) in commit [`178de00`](https://github.com/ukgovdatascience/govcookiecutter/pull/31/commits/178de0063a87f15b35fed7fbb8052cb81d12dbf9). Will do an [accessibility test similar to GOV.UK Developer Docs](https://docs.publishing.service.gov.uk/accessibility.html#preparation-of-this-accessibility-statement) at a later date!

File: README.md [hunk 6]
```diff
@@ -80,49 +91,52 @@ For Debian/Ubuntu, use the following commands:
...
+- Want to add some project-specific checklists to the pull/merge request template? See
+  the relevant Markdown files within the `.github` (GitHub) or
+  `.gitlab/merge_request_templates` (GitLab) folder
 
 ## Licence
 
-Unless stated otherwise, the codebase is released under the MIT License. This covers both the codebase and any sample
-code in the documentation. The documentation is © Crown copyright and available under the terms of the Open Government
-3.0 licence.
+Unless stated otherwise, the codebase is released under the MIT License. This covers
```
[the remark this answers is not in the document]
Comment (ESKYoung) [reply README.md:129]: Based this paragraph and the licence choice on other alphagov repositories, for example [govuk-frontend's licence](https://github.com/alphagov/govuk-frontend#licence), and also the [Service Manual](https://www.gov.uk/service-manual/technology/making-source-code-open-and-reusable#licensing-your-code).

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -1,30 +1,31 @@
...
 
-Add your summary here - keep it brief, to the point, and in plain English. For further information about pull requests,
-check out the [GDS Way][gds-way].
+Add your summary here - keep it brief, to the point, and in plain English. For further
+information about pull requests, check out the
+[GDS Way](https://gds-way.cloudapps.digital/standards/pull-requests.html).
 
 # Checklists
 
 <!--
```
[the remark this answers is not in the document]
Comment (ESKYoung) [reply .github/pull_request_template.md:9]: The PR template is designed to nudge contributors to raise PRs consistently; the the outputted projects, this includes an AQA checklist (but not here). The HTML comments provide some additional context for the contributors that doesn't need to be seen by others (unless they are editing the PR summary themselves).

File: hooks/post_gen_project.py (hunk 2 above)
Comment (ESKYoung) [reply hooks/post_gen_project.py:17]: To confirm, with a Standard Windows account, I can run `govcookiecutter`, and this function does not raise any errors, and the files are deleted as expected. Note I ran this within the `Documents` folder, so [creating a new project in a folder that requires user access control](https://docs.microsoft.com/en-us/windows/security/identity-protection/user-account-control/how-user-account-control-works) may fail.

File: docs/reference/README.md [hunk 8]
```diff
@@ -1,10 +1,41 @@
...
+```
+
+## Post-generation clean up
+
+```{eval-rst}
+.. autosummary::
+    :toctree: api/
+
+    delete_files_and_folders
+    parse_features_json
```
Comment (exfalsoquodlibet) [inline docs/reference/README.md:39]: Just for understanding, what do these functions do?

**Comment by exfalsoquodlibet** [comment]
Great additions and changes @ESKYoung, looks really good to me. 

I tested it both as by building a template and review the documentation as a "user", and by cloning govcookiecutter and building and reviewing the documentation as a "contributor". All's good.

File: .github/pull_request_template.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (maxf) [reply .github/pull_request_template.md:9]: understood

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxf) [reply README.md:129]: That's ok. It's a point I always raise because I'm not sure why we don't use OGL more. Probably historical

File: hooks/post_gen_project.py (hunk 2 above)
Comment (maxf) [reply hooks/post_gen_project.py:17]: ok

**Comment by ESKYoung** [comment]
Thanks @maxf and @exfalsoquodlibet for your review and comments! Approved offline, so will merge in now!

**Comment by maxf** [comment]
:clap:

**Comment by MarcoGorelli** [comment]
> Ah thanks for the tip @MarcoGorelli! And thanks for the package - it's really useful for what we're trying to achieve here! tada

Happy to help, glad you've found it useful!

> Yes I believe the celltags solution would work

@ESKYoung as of 0.13.0, skipping cells based on celltags should be available - would really appreciate it if you could try this out and report any bugs, thanks :+1: 
And feel free to reach out if you encounter any other issues
